runtime/lua/vim/diagnostic.lua

```diff
@@ -1857,16 +1857,37 @@ function M.open_float(opts, ...)
   if scope == 'line' then
     --- @param d vim.Diagnostic
     diagnostics = vim.tbl_filter(function(d)
-      return lnum >= d.lnum and lnum <= d.end_lnum
+      if lnum < d.lnum or lnum > d.end_lnum then
+        return false
+      end
+      if d.lnum == d.end_lnum then
+        -- if diagnostic only spans one line, always show
+        -- allows showing diagnostics with zero width (where lnum == end_lnum and col == end_col)
+        return true
+      else
+        -- end-exclusive range, don't show if on end_lnum and end_col == 0
+        return lnum ~= d.end_lnum or d.end_col ~= 0
+      end
     end, diagnostics)
   elseif scope == 'cursor' then
-    -- LSP servers can send diagnostics with `end_col` past the length of the line
-    local line_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
+    -- LSP servers can send diagnostics with `col` past the end of the line
+    local lnum_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
     --- @param d vim.Diagnostic
     diagnostics = vim.tbl_filter(function(d)
-      return d.lnum == lnum
-        and math.min(d.col, line_length - 1) <= col
-        and (d.end_col >= col or d.end_lnum > lnum)
+      if lnum < d.lnum or lnum > d.end_lnum then
+        return false
+      end
+      if lnum == d.lnum and col < math.min(d.col, lnum_length - 1) then
+        return false
+      end
+      if d.lnum == d.end_lnum and d.col == d.end_col then
+        -- treat zero-width diagnostics as having width 1
+        -- condition is effectively this:
+        -- return lnum ~= d.end_lnum or col < (d.col + 1)
+        return col == d.col
+      else
+        return lnum ~= d.end_lnum or col < d.end_col
+      end
```

Comment:
I'm realizing that we already have all of this logic in our `_range` module, but that is currently namespaced under `vim.treesitter` so wouldn't be appropriate to use here. A follow up PR would be to move `_range.lua` into a higher-level/more generic namespace, and refactor some of the diagnostics code to use that instead.